### PR TITLE
Prevent segfault upon running out of addresses

### PR DESCRIPTION
**Pull request by tiyash-basu-sociomantic**
Previously, `addrinfo_next` was not checked to be null before de-referencing to get `ai_next`. This would often result in segfaults.

In this PR, upon running out of addresses (i.e., `addrinfo_next == nullptr`), `getaddrinfo` is be called again (via `drizzle_state_addrinfo`) to refresh the list of addresses.

### CODE REVIEW

**Comment by tiyash-basu-sociomantic** [comment]
Updated to return `DRIZZLE_RETURN_COULD_NOT_CONNECT` while reading when `addrinfo_next == nullptr`

File: src/conn.cc [hunk 1]
```diff
@@ -1765,6 +1777,10 @@ static void connect_failed_try_next(drizzle_st *con, const char *file, uint line
   drizzle_set_error(con, file, line, function, "connect: %s (port %s): %s",
                     hostbuf, servbuf, msg);
 
-  con->addrinfo_next= con->addrinfo_next->ai_next;
+  if (con->addrinfo_next != nullptr)
```
Comment (ben-palmer-sociomantic) [inline src/conn.cc:1780]: Shouldn't you also be returning `DRIZZLE_RETURN_COULD_NOT_CONNECT` if this fails?

Comment (andreas-bok-sociomantic) [reply src/conn.cc:1780]: I don't reckon so. But I did a modification of the PR where this is added : 
```c++
static void connect_failed_try_next(drizzle_st *con, const char *file, uint line,
  const char *function, const char *msg)
{
  ...

  if (con->addrinfo_next == nullptr)
  {
    drizzle_log_info(con, __FILE_LINE_FUNC__, "refreshing addresses");
    con->push_state(drizzle_state_connect);
    con->push_state(drizzle_state_addrinfo);
  }
  else
  {
    con->addrinfo_next= con->addrinfo_next->ai_next;
    con->push_state(drizzle_state_connect);
  }
}
```

if the addrinfo is NULL `drizzle_state_addrinfo` would try to refresh it before `drizzle_state_connect` is entered again. If `drizzle_state_connect` connects we would with the changes return **DRIZZLE_RETURN_COULD_NOT_CONNECT** to the client

Comment (tiyash-basu-sociomantic) [reply src/conn.cc:1780]: Ah, right. This should have been left as it was initially in this PR.

**Comment by andreas-bok-sociomantic** [comment]
Closed as problem was resolved at the user's side

**Comment by tiyash-basu-sociomantic** [comment]
> Closed as problem was resolved at the user's side

This is not resolved with [redacted] yet.

**Comment by bokchan** [comment]
@tiyash-basu-sociomantic, did you get more details about this from your side?

**Comment by tiyash-basu-sociomantic** [comment]
> @tiyash-basu-sociomantic, did you get more details about this from your side?

I have been using this fix in my tests and have not encountered any segfault so far.

**Comment by bokchan** [comment]
@tiyash-basu-sociomantic 
> I have been using this fix in my tests and have not encountered any segfault so far.

Great. I suggest that we do a `v6.1.6` patch release as soon as possible. Then subsequently `v6.2.0` should soon be ready for release

**Comment by tiyash-basu-sociomantic** [comment]
> Great. I suggest that we do a `v6.1.6` patch release as soon as possible. Then subsequently `v6.2.0` should soon be ready for release

Absolutely! Please let me know if can help with anything.

**Comment by bokchan** [comment]
@tiyash-basu-sociomantic could you rebase on `v6.1.x`?

**Comment by tiyash-basu-sociomantic** [comment]
> @tiyash-basu-sociomantic could you rebase on `v6.1.x`?

Sure! I'll change the target branch and rebase (the target branch is `v6.x.x` atm).

Edit: Done!

**Comment by bokchan** [comment]
@tiyash-basu-sociomantic try to rebase locally on the updated `v6.1.x` and `force push`

**Comment by tiyash-basu-sociomantic** [comment]
> @tiyash-basu-sociomantic try to rebase locally on the updated `v6.1.x` and `force push`

It's already on `v.6.1.x` branch, and the diff confirms that. Not sure why this failure happens.
```
libdrizzle-redux-container | make: *** No rule to make target 'api-sanity-checker'.  Stop.
libdrizzle-redux-container exited with code 2
```

**Comment by bokchan** [comment]
> It's already on `v.6.1.x` branch, and the diff confirms that. Not sure why this failure happens.

No a bit strange. I tried to clear the build cache on travis.org and manually start a new build. Hopefully that solves the issue

**Comment by bokchan** [comment]
Nope. travis is for some reason still using the config file with the api-sanity-checker build. Could you try to do a force-push, just some random whitespace change, to force a new build on travis?

**Comment by bokchan** [comment]
Yep. Now it is working. Just revert if you did any changes and it is good to go :)

**Comment by tiyash-basu-sociomantic** [comment]
> Yep. Now it is working. Just revert if you did any changes and it is good to go :)

Nice! :) I just made a harmless updated to a comment:
`Try the next address, if available.` => `Try the next address, if it is available.`

**Comment by tiyash-basu-sociomantic** [comment]
Thanks!

**Pull request merged**